(closes #1584, #1041, #490) Add OMP/ACC directives to support device subroutines

sergisiso opened the pull request:
Closes #1584 , #1041, #490

sergisiso commented:
@arporter This is ready for the first review. This PRs creates the OpenMP and OpenACC directive and the OpenMP transformation to add it. I then realised that an ACCRoutineTrans already existed but is not used in Nemo (I guess it is for kernels in GOcean and LFRic?). This existing transformation modifies the fparser tree directly. For the moment I left it untouched since it was starting to diverge to what I am interested in for NEMO, but if it is just a matter of replace its contents I could do it in this PR. If you foresee other complications I will leave it for another issue.

arporter commented:
This merging/updating of ACCRoutineTrans is the subject of #1041. Strangely, I see that the transformation does already update the PSyIR of the kernel (as well as the parse tree) but that it uses a CodeBlock rather than a Directive node. It may be that all we need to do is update that bit and remove the bit that modifies the parse tree.

arporter commented on src/psyclone/psyir/nodes/acc_directives.py:
```diff
@@ -101,6 +101,36 @@ class ACCStandaloneDirective(ACCDirective, StandaloneDirective):
+        # Generate the code for this Directive
+        parent.add(DirectiveGen(parent, "acc", "begin", "routine", ""))
+
+    def begin_string(self):
```
Could we replace this with the new Clause support now?

arporter replied:
Ideally I think we may want to support clauses such as `seq` and `vector` but I'll check with the various project members as to whether we have an immediate requirement.

sergisiso replied:
I have not added clauses yet as it seems al current cases will do with the standard directive.

That is if I understand you correctly, when adding clauses we will still use the begin_string as it is now. It will be the backend that will concatenate the begin_string with the result of visiting the children clauses.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -120,6 +120,38 @@ class OMPStandaloneDirective(OMPDirective, StandaloneDirective):
+        # Generate the code for this Directive
+        parent.add(DirectiveGen(parent, "omp", "begin", "declare", "target"))
+
+    def begin_string(self):
```
Again, could we give this Node Clauses instead?

sergisiso replied:
I am not sure about this one. Declare doesn't exist alone, and in fact the OpenMP standard speaks about the "declare target directive" with the clauses being "to" "link" and "device_type" https://www.openmp.org/spec-html/5.0/openmpsu62.html

arporter replied:
It's OK. I was thinking that with Clause nodes we would do away with `begin_string` altogether but of course we don't.

arporter commented on src/psyclone/psyir/nodes/acc_directives.py:
```diff
@@ -101,6 +101,36 @@ class ACCStandaloneDirective(ACCDirective, StandaloneDirective):
     ''' Base class for all standalone OpenACC directive statements. '''
 
 
+class ACCRoutineDirective(ACCStandaloneDirective):
```
Copyright date please.

sergisiso replied:
Done

arporter commented on src/psyclone/f2pygen.py:
```diff
@@ -141,7 +141,7 @@ class OMPDirective(Directive):
     def __init__(self, root, line, position, dir_type):
         self._types = ["parallel do", "parallel", "do", "master", "single",
-                       "taskloop", "taskwait"]
+                       "taskloop", "taskwait", "declare"]
```
Copyright please.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/nodes/acc_directives_test.py:
```diff
@@ -242,7 +243,19 @@ def test_acckernelsdirective_gencode(default_present):
         "      !$acc end kernels\n" in code)
 
-# Class ACCKernelsDirective end
+
```
Since we're here would you mind changing lines 240-241 to use an f-string since that's the only one in this file?

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -641,7 +641,81 @@ def apply(self, node, options=None):
+    ...     end subroutine
+    ...     """
+    >>> omptargettrans = OMPDeclareTargetTrans()
+    >>> omptargettrans.apply(tree.walk(Routine))
```
Should that be `tree.walk(Routine)[0]`?

sergisiso replied:
Indeed. Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -641,7 +641,81 @@ def apply(self, node, options=None):
+        :param node: the PSyIR node to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
s/dictionary of string/dict of str/

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -274,6 +275,36 @@ def test_omptargettrans(sample_psyir):
     assert loops[0].parent.parent is loops[1].parent.parent
 
 
+def test_ompdeclaretargettrans(sample_psyir, fortran_writer):
```
Since we're here and it's just the one, please could you change line 214 to use an f-string.

sergisiso replied:
Done

arporter reviewed (changes requested):
Looks good Sergi, I have only very minor comments.
I think it would be valuable to at least try and see if the existing ACCRoutineTrans can be updated without too much trouble. If it gets complicated then we can do it as part of the issue I referred to earlier.
I've checked everything with pylint.
I'll check compilation and docs next time. In fact, now that I write that - please could you add the new transformation to the user guide.

sergisiso commented:
@arporter Changing the OpenACC implementation from fparser modification to using PSyIR was easy enough, so I added some more code compared to the first review, but this now should close 3 related issues. This is ready for another review.

arporter commented on doc/user_guide/transformations.rst:
```diff
@@ -755,6 +755,7 @@ transformations currently supported allow the addition of:
 
 * an **OpenMP Parallel** directive
 * an **OpenMP Target** directive
+* an **OpenMP Declare Target** directive
```
Please could you also update the AvailableTransformations section with a `.. autoclass` directive for this new transformation (in alphabetical order with all the others).

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2840,101 +2863,89 @@ def name(self):
-        This transformation affects the f2pygen and the PSyIR trees of
-        this kernel.
+        Add the '!$acc routine' OpenACC directive into the code of the
+        supplied kernel or routine.
```
I think we need some words here to explain why we distinguish kernel and routine. Perhaps "... kernel (in a PSyKAl API such as GOcean or LFRic) or PSyIR Routine."

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2840,101 +2863,89 @@ def name(self):
-        kernel_schedule = kern.get_kernel_schedule()
-        kernel_schedule.addchild(
-            CodeBlock([cmt], CodeBlock.Structure.STATEMENT), 0)
+        # Insert the directive to the routine if it doesn't already exist
```
This is so much nicer :-)
Contrary to what I said in an earlier comment, I think we can remove that test that the directive gets added at the end of the declarations - it's not an issue anymore.

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2840,101 +2863,89 @@ def name(self):
+                kernel_schedule = node.get_kernel_schedule()
+            except GenerationError as error:
+                raise TransformationError(
+                    f"Failed to create PSyIR version of kernel code for "
```
s/version//

arporter replied:
Also, please could you extend the comment to explain why this might fail (i.e. the reason for the try...except).

sergisiso replied:
Done both

arporter reviewed (changes requested):
Looking good now - it's really pleasing to see that old transformation code replaced.
Just a few minor things to tidy up.
All examples and tests are OK with compilation. (I see coverage is up to 99.33% now.)
All docs build OK.
Issue #490 is referenced:

    ./src/psyclone/psyGen.py:1898:                # exist - #490.

but none of the others are.

sergisiso commented on examples/nemo/scripts/kernels_trans.py:
```diff
@@ -662,6 +663,14 @@ def trans(psy):
+        # In the lib_fortran file we annotate each routine that does not
+        # have a Loop or a Call with the OpenACC Routine Directive
+        if psy.name == "psy_lib_fortran_psy":
+            if not sched.walk(Loop) and not sched.walk(Call):
```
Done

sergisiso commented on src/psyclone/psyir/nodes/codeblock.py:
```diff
@@ -132,5 +134,13 @@ def node_str(self, colour=True):
+        '''
+        parse_tree = self.get_ast_nodes
+        return [node.string for node in walk(parse_tree, Fortran2003.Name)]
+
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -151,6 +152,24 @@ def begin_string(self):
+        if self.parent and (not isinstance(self.parent, Routine) or
+                            self.parent.children[0] is not self):
+            raise GenerationError(
+                f"A OMPDeclareTargetDirective can only be found as a child 0 "
```
Done

sergisiso commented on src/psyclone/tests/psyir/nodes/codeblock_test.py:
```diff
@@ -40,13 +40,14 @@
 
 from __future__ import absolute_import
 import pytest
+from fparser.common.readfortran import FortranStringReader
```
Done

sergisiso commented on src/psyclone/tests/psyir/nodes/codeblock_test.py:
```diff
@@ -99,3 +100,19 @@ def test_assignment_children_validation():
+    end subroutine mytest''')
+    prog = parser(reader)
+    block = CodeBlock(prog.children, CodeBlock.Structure.STATEMENT)
+    assert "a" in block.get_symbol_names()
```
I added one in the test sample code and it is not being returned as a name, which is good.

sergisiso commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -301,8 +302,55 @@ def test_ompdeclaretargettrans(sample_psyir, fortran_writer):
+        not_declared1 = not_declared1 + not_declared2
+    end subroutine mytest''')
+    prog = parser(reader)
+    block = CodeBlock(prog.children[0].content[1].content[0].items,
```
Done

sergisiso commented on src/psyclone/transformations.py:
```diff
@@ -717,6 +710,36 @@ def validate(self, node, options=None):
+            for name in names:
+                try:
+                    candidate.scope.symbol_table.lookup(
+                        name, scope_limit=candidate.ancestor(nodes.Routine))
```
Good catch, done.

sergisiso commented on src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:
```diff
@@ -118,35 +115,43 @@ def test_accroutine_module_use():
 
 
-def test_accroutine_empty_kernel():
+def test_accroutine_empty_kernel(fortran_writer):
```
Removed as suggested by the comment below

sergisiso commented on src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:
```diff
@@ -118,35 +115,43 @@ def test_accroutine_module_use():
 
 
-def test_accroutine():
+def test_accroutine_to_kern(fortran_writer):
```
I moved this and the other 2 similar ACCRoutineTrans tests as suggested. I appreciate that this doesn't follow the "test in test/+ same location as code being tested" rule.
I started trying to move the OpenACC routines out of src/psyclone/transformations.py and the same for the associated tests but each step had many chain effects, broken imports, broken transinfo, ... So I decided to not do it in this PR, but is something that I may attempt in a separate PR.

sergisiso commented:
@arporter This is ready for another look:

> Issue https://github.com/stfc/PSyclone/issues/490 is referenced:
> ./src/psyclone/psyGen.py:1898:                # exist - #490.

Apparently this was the last transformation that modified the ast directly from what I can gather. Removing the _rename_ast didn't make any test or example fail. So I removed this functionality and updated the developer's guide.

arporter commented on doc/developer_guide/transformations.rst:
```diff
@@ -55,43 +55,21 @@ Transformations
-to generate the PSyIR representation of the kernel code. 
+PSyclone is able to perform kernel transformations by obtaining the PSyIR
+representation of the kernel with:
 
```
This is great but the docstring for `get_kernel_schedule` still talks about the fp2 AST. Please could you update it.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/nodes/codeblock.py:
```diff
@@ -132,5 +134,13 @@ def node_str(self, colour=True):
+        :returns: the list of symbol names used inside the CodeBock.
+        :rtype: list of str
+        '''
+        parse_tree = self.get_ast_nodes
```
Thanks.

arporter commented on src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:
```diff
@@ -32,8 +32,9 @@
 # Author: A. R. Porter, STFC Daresbury Lab
-# Modified by: R. W. Ford, STFC Daresbury Lab,
+# Modified by: R. W. Ford, STFC Daresbury Lab
 #              I. Kavcic, Met Office.
```
It's a tiny thing but it bothers me so please could you remove the full stop from this line.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:
```diff
@@ -460,13 +367,23 @@ def test_kernel_trans_validate(monkeypatch):
+
+    def raise_gen_error():
+        '''Simple function that raises GenerationError.'''
+        from psyclone.errors import GenerationError
```
Please could you move this import (and the one earlier for SymbolError) up to the top to improve pylint a bit.

arporter replied:
Actually, I realise now that pylint says that `GenerationError` has already been imported.

sergisiso replied:
Moved all imports to the top and ordered them alphabetically

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2925,13 +2924,16 @@ def validate(self, node, options=None):
                     f"'{type(node).__name__}'")
 
-            # Get the PSyIR routine from the associated kernel
+            # Get the PSyIR routine from the associated kernel If there is an
```
"...kernel. If there..."

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2925,13 +2924,16 @@ def validate(self, node, options=None):
             try:
                 kernel_schedule = node.get_kernel_schedule()
-            except GenerationError as error:
+            except Exception as error:
```
Is this form OK (was it a bare `except:` that you were saying we should avoid using the other day)?

sergisiso replied:
Yes, the bare `except:` includes things like keyboard or OS interrupts and should not be used. But these do not subclass Exception, so Exception is safe to use.

arporter reviewed (commented):
Very, very nearly there now.
Tests and examples are all OK with compilation.
Docs have been updated and build and look (almost) fine.
Please could you bring up-to-date with master and address the few minor comments inline.

sergisiso commented:
@arporter Ready for another look

arporter reviewed (approved):
All requested changes have been made.
Dev. Guide is now good. All docs build fine.
All tests and examples are OK with compilation.
